src/tainting/Function_call_graph.ml

```diff
@@ -577,160 +579,110 @@ let identify_callback ?(all_funcs = []) ?(caller_parent_path = [])
-                (* First try to find as class method *)
-                let current_class = match caller_parent_path with
-                  | Some cls :: _ -> Some cls
-                  | _ -> None
-                in
-                let hof_match = match current_class with
-                  | Some cls ->
-                      let class_name_str = fst cls.IL.ident in
-                      List.find_opt (fun (fn_id, _) ->
-                        match fn_id with
-                        | [Some c; Some m] when fst c.IL.ident = class_name_str && fst m.IL.ident = callee_name_str -> true
-                        | _ -> false
-                      ) user_hofs
-                  | None ->
-                      (* Try as free function *)
-                      List.find_opt (fun (fn_id, _) ->
-                        match fn_id with
-                        | [None; Some m] when fst m.IL.ident = callee_name_str -> true
-                        | _ -> false
-                      ) user_hofs
-                in
-                (match hof_match with
-                | Some (_, called_params) ->
-                    (* Extract callbacks from the arguments at the indices where parameters are called *)
-                    let arg_list = Tok.unbracket args in
-                    List.iter (fun (_param_name, param_idx) ->
-                      match List.nth_opt arg_list param_idx with
-                      | Some (G.Arg arg_expr) ->
-                          (match extract_callback_from_arg arg_expr with
-                          | Some callback_name ->
-                              (match identify_callback ~all_funcs ~caller_parent_path callback_name with
-                              | Some fn_id ->
-                                  Log.debug (fun m -> m "USER_HOF_CALLBACK: Found callback %s at arg %d in call to %s"
-                                    (fst callback_name.IL.ident) param_idx func_name);
-                                  (* Use fake token with callback name - HOF edges shouldn't match regular call lookups *)
-                                  let hof_tok = Tok.unsafe_fake_tok (Printf.sprintf "<hof_callback:%s>" (fst callback_name.IL.ident)) in
-                                  callbacks := (fn_id, hof_tok) :: !callbacks
-                              | None -> ())
-                          | None -> ())
-                      | _ -> ()
-                    ) called_params;
-                    super#visit_arguments env args
-                | None -> super#visit_expr env e))
-
-        | _ -> super#visit_expr env e
+        (match e.G.e with
+        | G.Call (callee, args) ->
+            let found = extract_hof_callbacks_from_call
+              ~method_hofs ~function_hofs ~user_hofs ~all_funcs ~caller_parent_path
+              callee args
+            in
+            callbacks := found @ !callbacks
+        | _ -> ());
+        super#visit_expr env e
```

Comment:
Leaving as TODO to extract this into a class, there are some potential concurrency issues with objects + domains.
It's one of the reasons I converted most (all?) such objects into classes and had the instances stateless & shared.